File: backend/app/api/endpoints/decisions.py
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, Query, Path
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.db.repositories import decision_repository
from app.schemas.decision import (
    Decision,
    DecisionCreate,
    DecisionList,
    DecisionUpdate,
    DecisionType,
)

router = APIRouter()


@router.get("/{decision_id}", response_model=Decision)
async def get_decision(
    decision_id: int = Path(..., title="The ID of the decision to get"),
    db: Session = Depends(get_db),
):
    """Get a decision by ID"""
    decision = decision_repository.get_by_id(db, decision_id)
    if not decision:
        raise HTTPException(status_code=404, detail="Decision not found")
    return decision


@router.get("/", response_model=DecisionList)
async def list_decisions(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=100),
    decision_type: Optional[DecisionType] = Query(None),
    db: Session = Depends(get_db),
):
    """List decisions with optional filtering"""
    decisions = decision_repository.get_all(
        db, skip=skip, limit=limit, decision_type=decision_type
    )
    total = decision_repository.count(db, decision_type=decision_type)
    return {"items": decisions, "total": total}


@router.post("/", response_model=Decision, status_code=201)
async def create_decision(
    decision_create: DecisionCreate,
    db: Session = Depends(get_db),
):
    """Create a new decision"""
    return decision_repository.create(db, decision_create)


@router.put("/{decision_id}", response_model=Decision)
async def update_decision(
    decision_update: DecisionUpdate,
    decision_id: int = Path(..., title="The ID of the decision to update"),
    db: Session = Depends(get_db),
):
    """Update an existing decision"""
    updated_decision = decision_repository.update(db, decision_id, decision_update)
    if not updated_decision:
        raise HTTPException(status_code=404, detail="Decision not found")
    return updated_decision


@router.delete("/{decision_id}", status_code=204)
async def delete_decision(
    decision_id: int = Path(..., title="The ID of the decision to delete"),
    db: Session = Depends(get_db),
):
    """Delete a decision"""
    success = decision_repository.delete(db, decision_id)
    if not success:
        raise HTTPException(status_code=404, detail="Decision not found")
    return None

File: frontend/src/services/decisions.ts
import axios from 'axios';
import { Decision, DecisionCreate, DecisionUpdate, DecisionType } from '../types';

// API base URL
const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000/api/v1';

// Decision API endpoints
const DECISIONS_API = {
  base: `${API_BASE_URL}/decisions`,
  byId: (id: string) => `${API_BASE_URL}/decisions/${id}`,
};

// Create axios instance
const apiClient = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Decision service
export const decisionService = {
  /**
   * Get all decisions with optional filtering
   */
  getDecisions: async (
    page = 1,
    limit = 10,
    type?: DecisionType
  ): Promise<{ items: Decision[]; total: number }> => {
    try {
      const skip = (page - 1) * limit;
      const params: Record<string, string | number> = { skip, limit };

      if (type) {
        params.decision_type = type;
      }

      const response = await apiClient.get(DECISIONS_API.base, { params });
      return response.data;
    } catch (error) {
      console.error('Error getting decisions:', error);
      throw error;
    }
  },

  /**
   * Get a decision by ID
   */
  getDecision: async (id: string): Promise<Decision> => {
    try {
      const response = await apiClient.get(DECISIONS_API.byId(id));
      return response.data;
    } catch (error) {
      console.error(`Error getting decision ${id}:`, error);
      throw error;
    }
  },

  /**
   * Create a new decision
   */
  createDecision: async (decision: DecisionCreate): Promise<Decision> => {
    try {
      const response = await apiClient.post(DECISIONS_API.base, decision);
      return response.data;
    } catch (error) {
      console.error('Error creating decision:', error);
      throw error;
    }
  },

  /**
   * Update an existing decision
   */
  updateDecision: async (id: string, decision: DecisionUpdate): Promise<Decision> => {
    try {
      const response = await apiClient.put(DECISIONS_API.byId(id), decision);
      return response.data;
    } catch (error) {
      console.error(`Error updating decision ${id}:`, error);
      throw error;
    }
  },

  /**
   * Delete a decision
   */
  deleteDecision: async (id: string): Promise<void> => {
    try {
      await apiClient.delete(DECISIONS_API.byId(id));
    } catch (error) {
      console.error(`Error deleting decision ${id}:`, error);
      throw error;
    }
  },
};

export default decisionService;
